### DISCUSSION TITLE
multiple clients

### DISCUSSION THREAD

**Discussion by Manubi** [urql-graphql/urql#1523] [Q&A] [answered]
Hi,

I need to query two different graphql servers. One is `www.example.com/api/graphql/` the other one is `www.example.com/internal/api/graphql`. I know it's not the common way of doing things, but I need to implement it... And as I wanted to try urql I was wondering how I can do it.

I know it's possible with apollo to add a second client. Something like `ApolloClients { default: publicClient, private: privateClient}`

Is there an easy version for urql as well?

Thanks a lot!

**Comment by JoviDeCroock** [comment, marked as answer]
I'd suggest creating an abstraction over the hooks honestly, where you use the [`context`](https://formidable.com/open-source/urql/docs/api/core/#operationcontext) argument to our hooks to supply a different url. For insance:

```
const useMyCustomQuery = ({ internal, ...args }) => useQuery(...args, context: useMemo(() => ({ url: internal ? 'x' : 'y' }), [])
```

**Reply by Manubi** [reply]
Thanks for the quick reply. I am unfortunately using vue3.

So far I have this.

```
    const result = useCustomQuery({
      internal: true,
      query: NEWS_ARTICLES_QUERY
    });
```

```
export default function useCustomQuery({ internal, ...args }: any) {
  const context = { url: internal ? "/internal/api/graphql" : "/api/graphql" };
  return useQuery({ ...args, context });
}
```

Can you help me with the typings? Thanks!

**Reply by JoviDeCroock** [reply]
Something like:

```
import { useQuery, UseQueryArgs } from '@urql/vue';

export default function useCustomQuery<Data, Variables>(
  { internal, ...args }: UseQueryArgs<Data = any, Variables = object>
): UseQueryResponse<Data, Variables> {
  const context = { url: internal ? "/internal/api/graphql" : "/api/graphql" };
  return useQuery({ ...args, context });
}
```

**Reply by nandorojo** [reply]
@JoviDeCroock I notice you didn't memoize the `context` here. I assume this is just pseudocode for the types, but to clarify, will requests refetch when the `context` reference changes? (I would like them to, but just checking.)

**Reply by JoviDeCroock** [reply]
You are right to point this out, you should memoize it, then when url changes it will refetch yes. In your case:

```
const context = useMemo(() => ({ url: internal ? "/internal/api/graphql" : "/api/graphql" }), [internal]);
```

**Comment by ceefour** [comment]
Was discussed before in issue #429

**Comment by vilchesalves** [comment]
Hey urlq people. :)

I'd like to +1 this, with my quite extreme use case: we have multiple microservices served in AWS AppSync endpoints, and for top down architectural decisions we currently have 5 different graphql endpoints for the frontend to handle.

With Apollo we were initially importing and using clients locally, but recently we managed to split the link dynamically.

Is the approach here to just overwrite the top client url with a local url in the context of the useQuery?
Is it worth it to have the approach added to the documentation?

Cheers!

**Reply by kitten** [reply]
that's a workaround, hence it's also not recommended in the docs. Our stance here is and has been the same for a while, if you have multiple GraphQL endpoint that you all need to use in the same app, then you're likely not following best practices or using it for what it's good at 😅

So while we have a workaround for this, actually, we'd tell people to write an exchange to split their queries endpoints, but noone that we know of has done this yet. While this takes some investment in the form of maybe 10-20 lines of code, this removes the added logic from the UI/hooks to determine the URL

**Reply by Manubi** [reply]
Sorry to bother again. But how would you do it then if you would like to use urql with graphl-code-generator? Because the version from above doesn't really work for me as the useMutation is only taking a DocumentNode and no context.

**Reply by ceefour** [reply]
My plan is to use [Hasura](https://hasura.io/) that federates all different GraphQL (and non-GraphQL) backends & data sources.

Therefore, the client only connects to Hasura, and have a single federated GraphQL schema.

**Reply by Manubi** [reply]
Also if I use the useCustomQuery I loose all my types from the urql-graphql-code-generator plugin...

**Reply by Manubi** [reply]
> My plan is to use [Hasura](https://hasura.io/) that federates all different GraphQL (and non-GraphQL) backends & data sources.
> 
> Therefore, the client only connects to Hasura, and have a single federated GraphQL schema.

I already used Hasura. Thanks but no.

**Reply by kitten** [reply]
@manubi Ultimately, the long term solution is to write an exchange, maybe with a directive to add endpoints to the query.

On the topic of mutations, they do accept context; every hook/bindings support that in all parts, but it's a little "hidden" for mutations: https://github.com/FormidableLabs/urql/blob/7ec183808b2e766fbe1964c8d1ad6c8bc999d715/packages/vue-urql/src/useMutation.ts#L60

**Reply by ceefour** [reply]
@Manubi This is interesting. Can you share your experience why you moved away from Hasura, and approaching multiple GraphQL clients instead?

**Reply by JoviDeCroock** [reply]
@Manubi writing an exchange won't have you touching the custom-hooks. 

Also hooks look like this:

```
export function useFindUserQuery(options: Omit<Urql.UseQueryArgs<FindUserQueryVariables>, 'query'> = {}) {
  return Urql.useQuery<FindUserQuery>({ query: FindUserDocument, ...options });
};
```

Those options permit you to pass in the aforementioned context.

**Reply by Manubi** [reply]
> @Manubi This is interesting. Can you share your experience why you moved away from Hasura, and approaching multiple GraphQL clients instead?

I just moved away from Hasura because the bigger the projects gets the more cumbersome development with hasura gets. Also the onboarding of new devs just takes too long. Because everyone is afraid of touching something because nobody knows what's going on.

**Reply by Manubi** [reply]
Hey again,

I am looking into the exchange stuff which you can find [here](https://formidable.com/open-source/urql/docs/advanced/authoring-exchanges/). I wanted to pass a new context property with the client in order to read it in the exchange from the client and then change the URL if needed. But somehow I get an error when fetching the articles. 
<img width="771" alt="CleanShot 2021-07-07 at 15 59 26@2x" src="https://user-images.githubusercontent.com/20935138/124772483-5c081f00-df3c-11eb-89f7-bf7847caa148.png">

I want to read it from the client in the exchange:

```
export default function noopExchange({ client, forward }: any) {
  console.log("client", client.context.newUrl);
  console.log("forward", forward);
  return (operations$: any) => {
    // <-- The ExchangeIO function
    const operationResult$ = forward(operations$);
    return operationResult$;
  };
}

```

Anyone? thanks

**Reply by kitten** [reply]
@Manubi The client doesn't carry the context as it's more like an event hub or manager of operations. It only carries options with it. The input from your bindings (or calls to the Client) and the Client's base options combine into forming operations.

So when you're writing an exchange you're dealing with a stream of operations coming from the client or the last exchange in the chain, which you can then transform and `forward`. This stream then goes on since `forward` refers to the next exchange in your pipeline.

Each exchange then returns a stream of results, so that's what `forward` returns. In other words, we're piping a stream in and composing a stream back with results too. https://formidable.com/open-source/urql/docs/architecture/

Each `Operation` is unique and carries its own instance of a `context` on it, the kind of operation, a key, and the original query and variables https://formidable.com/open-source/urql/docs/api/core/#operation

If you're modifying the context in a binding it's hence different from operation to operation. This is important so that each operation _can_ actually have separate parameters.

You can actually see a simple example here: https://formidable.com/open-source/urql/docs/common-questions/#how-do-we-achieve-asynchronous-fetchoptions
But in practice, all you'll need is to `map` the incoming operations to change the context.

I'd still recommend you however to consider a directive on the query, e.g. `@api(url: "...")`, since this will not only ensure that your query's key will change, but since it'll also potentially allow you to integrate code/type generator tools more easily in the future with some custom plugins.

**Comment by Mookiies** [comment]
If using React a way to workaround the single client limit is by nesting another `Provider` with a different client. This works because [Providers can be nested to override values deeper within the tree](https://reactjs.org/docs/context.html#contextprovider).
  
This can work well without getting complicated so long as the nested parts of the tree only need to use one client.

**Reply by ross-rosario** [reply]
@Mookiies good idea! I wonder, how do you get to specify the client needed when calling `useQuery`?

**Reply by ross-rosario** [reply]
Gotcha! So a Provider is specific to a tree of components, as you suggested.

**Reply by Mookiies** [reply]
`useQuery` uses the client that it gets from the provider, which is determined by the closest `Provider` in the tree. So your usage of `useQuery` wouldn't specify anything, the client is determined by the react tree.

For example here `Component1` and `Component2` will use `client1`. However `Component3` will use `client2` because of the nested provider.

Index:
```
  <Provider value={client1}>
    <YourRoutes />
  </Provider>
```

YourRoutes
```
const client2 = createClient({});
...
<>
   <Component1 />
   <Component2 />
   <Provider value={client2}>
      <Component3 />
   </Provider>
</>
```

**Reply by Manubi** [reply]
Works like a charm! thanks :D

**Comment by tomduncalf** [comment]
The way I've solved this was inspired by what kitten wrote in https://github.com/FormidableLabs/urql/discussions/1523#discussioncomment-973843.

My use case is I am using a hosted CMS (DatoCMS) which has a GraphQL API, and also have a local API for e-commerce cart operations. I _could_ proxy the CMS API through my local API, but that makes no sense as it massively increases the amount of traffic the local API has to handle. 

I am using `graphql-codegen` to generate my hooks, so didn't want to have to manually mess with the hooks. Instead, I created an exchange which modifies the operation's URL and headers (as each API has a different authentication mechanism) based on the name of the selection set – i.e. the "thing" you are querying (lol – I'm new to GraphQL so not sure if there's a better term) – e.g. in this case, it would be `cart`:

```graphql
query GetCart {
  cart {
    id
    items {
      product {
        name
      }
    }
  }
}
```

I have an array of selection set names which should be sent to my local API, and the rest get sent to the CMS. This probably wouldn't work for more complex queries, but it works for my use case and could be expanded to use whatever details of `op` you care about, just log it out and see what is in there.

```javascript
const API_QUERY_SELECTION_SETS = ["cart", "clearCart"];

const urlSplitterExchange = ({ forward }) => {
  return (ops$) => {
    const shared = pipe(ops$, share);

    // Split out queries/mutations, which we want to handle...
    const queries = pipe(
      shared,
      filter((op) => op.kind === "query" || op.kind === "mutation"),
      onPush((op) => {
        // Get each operations first selection set and check if it's in API_QUERY_SELECTION_SETS
        // and set the URL and headers accordingly
        const selectionSet = op.query.definitions[0].selectionSet.selections[0].name.value;
        if (API_QUERY_SELECTION_SETS.includes(selectionSet)) {
          op.context.url = API_URL;
          op.context.fetchOptions.headers = API_HEADERS;
        } else {
          op.context.url = DATO_URL;
          op.context.fetchOptions.headers = DATO_HEADERS;
        }
      })
    );

    // ...and other messages, which we just forward
    const others = pipe(
      shared,
      filter((op) => !(op.kind === "query" || op.kind === "mutation"))
    );

    return forward(merge([queries, others]));
  };
};
```

Then add this exchange *before* the `fetchExchange`.

So with the above example, queries or mutations named `cart` or `clearCart` go to the local API, everything else goes to the CMS. Hope this helps someone!

Cheers,
Tom

**Comment by ross-rosario** [comment]
The problem with changing client's URL on the fly, is that cache is unintentionally shared. If you send a query to Url1, then for whatever reason, send the same query to Url2, you'll hit the cached response of Url1.

**Reply by tomduncalf** [reply]
True, it would be nice if there were a better way to work around this

**Comment by agcty** [comment]
I'm using urql to fetch some data from the graph (graphql for ethereum basically). I have a similar issue with using different URLs, I need to fetch a lot of different data from different graph endpoints and desperately need a nice way of using different urls.

**Reply by JoviDeCroock** [reply]
Hey, you can leverage your own exchange for this

```ts
import { dedupExchange, cacheExchange, fetchExchange, createClient } from 'urql';
import {
  makeOperation,
  Operation,
  OperationResult,
  Exchange,
} from '@urql/core';
import { pipe, tap, map } from 'wonka';

// You can use options for inputs if you want to make this reusable for  the community
export const requestPolicyExchange = (options): Exchange => ({
  forward,
}) => {
  const processIncomingOperation = (operation: Operation): Operation => {
    if (operation) {
      // your own condition
      return makeOperation(operation.kind, operation, {
        ...operation.context,
        url: 'some-url',
      });
    }
    
    return operation
  };

  return ops$ => {
    return pipe(
      map(processOperation),
      forward(ops$),
    );
  };
};

const client = createClient({
  url: 'some-url',
  exchanges: [dedupExchange, cacheExchange, myExchange(), fetchExchange]
})
```

**Reply by agcty** [reply]
With this code, do I still need different providers for caching? 🤔

**Reply by JoviDeCroock** [reply]
This uses 1 cache for all data, from my knowledge of the graph that shouldn't be an issue 😅 but my knowledge is quite limited in all honesty. We currently don't have a url-segmented cache but in essence it should be possible to implement

**Reply by mwilc0x** [reply]
Thank you for this helpful example!

I had to make a small change to get this working in my code, from the [docs](https://formidable.com/open-source/urql/docs/advanced/authoring-exchanges/#only-one-operations-stream). In case anyone new sees this!

```js
return ops$ => {
  return pipe(
    ops$,
    map(processIncomingOperation),
    forward,
  );
};
```

**Reply by AjayiMike** [reply]
Unfortunatelly, this approach doesn't work for me.

seems like `requestPolicyExchange` held on to the first context. it doen't update to change the endpoint in subsequent calls.

This is my code.

```typescript
import { ReactNode } from 'react';
import {
    cacheExchange,
    Client,
    Exchange,
    fetchExchange,
    makeOperation,
    Operation,
    Provider as UrqlProvider,
} from 'urql';
import { authExchange, AuthConfig } from '@urql/exchange-auth';
import { SUBGRAPH_ENDPOINTS } from '@/constants/urls';
import { DEFAULT_CHAIN_ID } from '@/constants/chains';
import { map, pipe } from 'wonka';

interface IAppUrqlProvider {
    children: ReactNode;
}

export interface Options {
    endpoints: {
        [chainId: string]: string;
    };
}

const options: Options = {
    endpoints: SUBGRAPH_ENDPOINTS,
};

const AppUrqlProvider: React.FC<IAppUrqlProvider> = ({ children }) => {
    const client = new Client({
        url: SUBGRAPH_ENDPOINTS[DEFAULT_CHAIN_ID],
        exchanges: [
            cacheExchange,
            authExchange(
                async (utils) =>
                    ({
                        addAuthToOperation: (operation) =>
                            utils.appendHeaders(operation, {
                                Authorization: `Basic ${btoa(
                                    `${process.env.NEXT_PUBLIC_SUBGRAPH_USERNAME}:${process.env.NEXT_PUBLIC_SUBGRAPH_PASSWORD}`
                                )}`,
                            }),
                    }) as AuthConfig
            ),
            requestPolicyExchange(options),
            fetchExchange,
        ],
    });
    return <UrqlProvider value={client}>{children}</UrqlProvider>;
};

export default AppUrqlProvider;

export const requestPolicyExchange =
    (options: Options): Exchange =>
    ({ forward }) => {
        const processIncomingOperation = (operation: Operation): Operation => {
            if (operation) {
                const endpoint =
                    options.endpoints[operation.context.clientNetworkId] ?? options.endpoints[DEFAULT_CHAIN_ID];

                return makeOperation(operation.kind, operation, {
                    ...operation.context,
                    url: endpoint,
                });
            }

            return operation;
        };

        return (ops$) => {
            return pipe(ops$, map(processIncomingOperation), forward);
        };
    };
```

How do I know?

the result.operation.context.networkClientId is not the same with networkClientId passed in the context when logged

The hook

```typescript
const useSubgraphQuery = (
        query: any,
        context: { [key: string]: any },
        variables: { [key: string]: any },
        requestPolicy = 'network-only' as RequestPolicy
    ) => {
        const [result] = useQuery({
            query,
            variables,
            context,
            requestPolicy,
        });

        console.log({ result, context });

        return result;
    };
```

**Comment by NotFounds** [comment]
Hello! Thank you for the awesome library.

Could you please reconsider https://github.com/urql-graphql/urql/pull/446.
We are using our own service's GraphQL API and a third-party provided GraphQL API on a single page. There are plenty of cases where a SaaS public API is exposed as a GraphQL API, so we believe that "GraphQL usually ties multiple APIs together and hence this doesn't really come up as we've seen." does not apply. (Examples of SaaS exposing GraphQL APIs: GitHub, Shopify, Stripe)

If using multiple SaaS APIs in a client is not recommended, what would be the best approach?

**Reply by kitten** [reply]
> There are plenty of cases where a SaaS public API is exposed as a GraphQL API

There are plenty of cases, yes, but all of the examples you've given just to make this clear, are APIs that you have to authenticate with.
Just to illustrate this, you wouldn't want your Stripe API tokens to be public and shared in a web application, right?

To take a step back, with GraphQL, at the end of the day we "just" build APIs. And there's nothing stopping us from accessing any API from our web app, and there's nothing even stopping you from adding multiple sub-sections to your app, or multiple `urql` clients, with some limitations.

However, when we speak about third-party APIs in your UI, we're speaking about losing basically all the benefits of GraphQL. Just because you're interacting with a GraphQL API, that doesn't mean it's _your_ GraphQL API, and you basically are losing benefits of schema design.

So, not to be rude, but it is a bit tongue in cheek 😅❤️, your example actually perfectly demonstrates why we don't even support this

**Reply by NotFounds** [reply]
Thank you for the quick reply.
You are right, the examples of GitHub and Stripe were not appropriate. My bad.

> you basically are losing benefits of schema design.

I understand.

However, Shopify seems to only provide a GraphQL API for the frontend. And, it appears that this API does not require authentication. If my understanding is correct, to use Shopify in a headless manner and build a UI, one must use the GraphQL API. By the way, there is a public token for API calls, so it seems there is no problem calling it from the browser.
https://shopify.dev/docs/api/storefront#authentication

Sorry for bringing up specific product examples. Should we integrate multiple APIs even in such cases?

**Reply by kitten** [reply]
> one must use the GraphQL API

Yes, but this is a service API. For example, say you're using an API with public resources. Even though you _could_ access its data from the front-end that doesn't mean that that's a good idea.

Say you've got a CMS, a shop API, and a payments API. For the sake of the argument, let's say they all have public APIs that _could_ be accessed from the front-end directly (i.e. CMS has a public GraphQL API for content, the shop has a public, unauthenticatd API for products, and the payment API has replicated products for the shop and payments can be initiated publicly by creating a cart via GraphQL).

Just because all three APIs _could_ be accessed from the frontend that doesn't mean it makes much sense in the context of `urql` and a GraphQL client. You'd be stuck having to stitch data to different data sources on the front-end, e.g. CMS says "list product A" then you look up the shop API for cart details "get product A by external ID", and then you create a cart entry via the payment API "for product A with external ID add it to cart". You basically end up doing something on the front-end that negates all strengths of GraphQL. At this point and with this usage, it basically doesn't matter that you're using GraphQL, since you've lost all benefits of it.

There's several workarounds listed in this thread, that all do allow you to do this, basically with a single `Client` in `urql`. So, you _can_ do this. But that doesn't mean it makes much sense.

The recommended way of using GraphQL, no matter whether you're consuming a third-party GraphQL API or any other kind of API, is to treat it as such — as a third-party API — and map this over to your own schema. If you don't want to go through the effort of this, you could look up "schema stitching", which can help with this by simply adding to your own schema with another GraphQL API, which a lot of utilities support (GraphQL Tools has some notes on this)

**Reply by NotFounds** [reply]
> You basically end up doing something on the front-end that negates all strengths of GraphQL. At this point and with this usage, it basically doesn't matter that you're using GraphQL, since you've lost all benefits of it.

Thank you for providing specific examples. I now understand the meaning of "But that doesn't mean it makes much sense."

> The recommended way of using GraphQL, no matter whether you're consuming a third-party GraphQL API or any other kind of API, is to treat it as such — as a third-party API — and map this over to your own schema. If you don't want to go through the effort of this, you could look up "schema stitching", which can help with this by simply adding to your own schema with another GraphQL API, which a lot of utilities support (GraphQL Tools has some notes on this)

I'll consider mapping the schema and also look into the concept of schema stitching. Thank you.
